src/rules/noAsyncWithoutAwaitRule.ts

```diff
@@ -0,0 +1,139 @@
+    }
+}
+
+class Walk extends Lint.RuleWalker {
+    protected visitFunctionDeclaration(node: ts.FunctionDeclaration) {
+        this.addFailureIfAsyncFunctionHasNoAwait(node);
+        super.visitFunctionDeclaration(node);
+    }
+
+    protected visitArrowFunction(node: ts.ArrowFunction) {
+        this.addFailureIfAsyncFunctionHasNoAwait(node);
+        super.visitArrowFunction(node);
+    }
+
+    protected visitMethodDeclaration(node: ts.MethodDeclaration) {
+        this.addFailureIfAsyncFunctionHasNoAwait(node);
+        super.visitMethodDeclaration(node);
+    }
+
+    protected visitFunctionExpression(node: ts.FunctionExpression) {
+        this.addFailureIfAsyncFunctionHasNoAwait(node);
+        super.visitFunctionExpression(node);
+    }
+
+    private getAsyncModifier(node: ts.Node) {
+        if (node.modifiers !== undefined) {
+            return node.modifiers.find((modifier) => modifier.kind === ts.SyntaxKind.AsyncKeyword);
+        }
+
+        return undefined;
+    }
+
+    private isAwait(node: ts.Node): boolean {
+        return node.kind === ts.SyntaxKind.AwaitKeyword;
+    }
+
+    private isReturn(node: ts.Node): boolean {
+        return node.kind === ts.SyntaxKind.ReturnKeyword;
+    }
+
+    private functionBlockHasAwait(node: ts.Node): boolean {
+        if (this.isAwait(node)) {
+            return true;
+        }
+
+        if (node.kind === ts.SyntaxKind.ArrowFunction
+            || node.kind === ts.SyntaxKind.FunctionDeclaration) {
+            return false;
+        }
+
+        // tslint:disable-next-line:no-unsafe-any
+        return node.getChildren().some(this.functionBlockHasAwait.bind(this));
+    }
+
+    private functionBlockHasReturn(node: ts.Node): boolean {
+        if (this.isReturn(node)) {
+            return true;
+        }
+
+        if (node.kind === ts.SyntaxKind.ArrowFunction
+            || node.kind === ts.SyntaxKind.FunctionDeclaration) {
+            return false;
+        }
+
+        const returnInChildren = node.getChildren()
+            .map((functionBlockNode: ts.Node) => this.functionBlockHasReturn(functionBlockNode));
+        return returnInChildren.some(Boolean);
+    }
+
+    private isShortArrowReturn(node: FunctionNodeType) {
+        return node.kind === ts.SyntaxKind.ArrowFunction && node.body.kind !== ts.SyntaxKind.Block;
+    }
+
+    private reportFailureIfAsyncFunction(node: FunctionNodeType) {
+        const asyncModifier = this.getAsyncModifier(node);
+        if (asyncModifier !== undefined) {
+            this.addFailureAt(asyncModifier.getStart(), asyncModifier.getEnd() - asyncModifier.getStart(), Rule.FAILURE_STRING);
+        }
+    }
+
+    private addFailureIfAsyncFunctionHasNoAwait(node: FunctionNodeType) {
+        if (node.body === undefined) {
+            this.reportFailureIfAsyncFunction(node);
+            return;
+        }
+        if (!this.isShortArrowReturn(node)
+            && !this.functionBlockHasAwait(node.body as ts.Node)
```

Comment:
> as `ts.Node`

These should be unnecessary because of the `if (node.body === undefined)` check, right?